compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6356,26 +6358,38 @@ private BType getCompletionType(BType collectionType, BLangQueryExpr queryExpr,
         return completionType;
     }
 
-    private BType getNonContextualQueryType(BType staticType, BType basicType) {
-        BType resultType;
+    private BType getNonContextualQueryType(BType staticType, BType basicType, Location pos) {
+        basicType = Types.getReferredType(basicType);
         switch (basicType.tag) {
             case TypeTags.TABLE:
-                resultType = symTable.tableType;
+                if (types.isAssignable(staticType, symTable.mapAllType)) {
+                    return symTable.tableType;
+                }
                 break;
             case TypeTags.STREAM:
-                resultType = symTable.streamType;
-                break;
+                return symTable.streamType;
+            case TypeTags.MAP:
+                dlog.error(pos, INCOMPATIBLE_QUERY_CONSTRUCT_MAP_TYPE);
+                return symTable.semanticError;
             case TypeTags.XML:
-                resultType = new BXMLType(staticType, null);
+                if (types.isSubTypeOfBaseType(staticType, symTable.xmlType.tag)) {
+                    return new BXMLType(staticType, null);
+                }
                 break;
             case TypeTags.STRING:
-                resultType = symTable.stringType;
+                if (types.isSubTypeOfBaseType(staticType, TypeTags.STRING)) {
+                    return symTable.stringType;
+                }
                 break;
+            case TypeTags.ARRAY:
+            case TypeTags.TUPLE:
+            case TypeTags.OBJECT:
```

Comment:
Can you please check how the below is validate
```ballerina
class Iterable {
    *object:Iterable;
    public function iterator() returns object {
        public isolated function next();
    } {
        return object {    
            public isolated function next()  {
            }
        };
    }
}

public function main() {
    Iterable ite = new();
    var listResult = from var item in ite
    select item;
}
```

Reply:
In here, the static type is become `other`.
It is happen because https://github.com/ballerina-platform/ballerina-lang/issues/38534#issuecomment-1301631231.
Created a issue for this https://github.com/ballerina-platform/ballerina-lang/issues/38534

Reply:
IMO, this should resolve in separate PR. @pcnfernando wdyt?